pkg/virtualization/core/virtualsystem/virtualmachine.go

```diff
@@ -772,22 +772,37 @@ func (vm *VirtualMachine) GetDvdDrives() (col drive.DvdDriveCollection, err erro
 	return settings.GetVirtualDvdDrives()
 }
 
-func (vm *VirtualMachine) GetDvdDriveAndLogicalDiskByIsoPath(isoPath string) (dvd *drive.DvdDrive, diskdvd *disk.LogicalDisk, err error) {
+func (vm *VirtualMachine) GetDvdDriveAndLogicalDiskByIsoPath(isoPath string) (dvdDrive *drive.DvdDrive, logicalDisk *disk.LogicalDisk, err error) {
 	dvdCol, err := vm.GetDvdDrives()
 	if err != nil {
 		return
 	}
 	defer dvdCol.Close()
 
-	for _, inst := range dvdCol {
-		dvddisk, err1 := drive.GetRelatedStorageAllocationSettingData(inst.WmiInstance)
+	// make sure the return values are closed and nil in case of error
+	defer func() {
+		if err != nil {
+			if dvdDrive != nil {
+				dvdDrive.Close()
+				dvdDrive = nil
```

Comment:
Do we need to set this to nil?

Reply:
I think this is just precautionary so that callers of this function will not call `Close()` again in `dvdDrive`. The callers of this function are expected to check if it's not `nil` before calling `Close()`  We wanted to make sure we returned nil for dvdDrive when there's an error.